First pass at adding Faker docs

prescod opened the pull request:
Includes a mkdocs plugin which generates localized documentation by executing Snowfakery on auto-generated snippets.

Basically the plugin introspects Faker to find all locales and fakers and then it synthesizes Snowfakery code to demonstrate the feature.

Other things in the PR:

 * Makefile for launching the whole process
 * Meta-docs about how to build the docs
 * Some doc reorganization (don't assume that "Green" text means new docs. Use Ctrl-F to see if it just moved)
 * Favicon
 * Docs updated to newest Faker syntax: first_name -> FirstName
  * Some Python code went in the opposite direction to be consistent with Faker's Python code. UserName->user_name (the customer can use either name in their recipe)
  * Fake data is classified as "common" or "uncommon"
  * Two files are noted as being copied and pasted from Faker. A detailed code review of them is probably not necessary (we wouldn't review it if we had included it via site-packages).
  * There is better hand-written documentation about how Fake data works in Snowfakery

davisagli commented:
@prescod Can you say a bit about the motivation for switching from sphinx to mkdocs? I am not familiar with mkdocs and don't want to use a different tool from our other projects unless there are compelling reasons.

davisagli commented:
I got an error when I tried to build the docs locally:
```
% make docs
python -m mkdocs build --clean --site-dir build/html --config-file mkdocs.yml
WARNING  -  Config value: 'theme_dir'. Warning: Unrecognised configuration name: theme_dir
ERROR    -  Config value: 'plugins'. Error: The "snowfakery_fakes" plugin is not installed
Aborted with 1 Configuration Errors!
make: *** [docs] Error 1
```

prescod commented:
David: it turns out that readthedocs.org was using MKDocs all along.

I looked at the build logs and realized it wasn't using Sphinx/myst. Which probably explains why I had no luck changing Sphinx themes.

Yes I can remove the rest of Sphinx.

I'll update the docs about the error you ran into. I do not know how to allow Mkdocs (or any other tool that uses setup.py Python entry points) to find plugins without "installing" the package. There is a whole extra file in the PR whose only purpose is to be the shim that serves as the entry point and then looks for the "real" plugin in the repo.

So you can either install or edit entrypoints.txt (which is what I did).

I would love to find a way that does not involve setup.py or editing that file but I don't know of one. It's a bit irritating, to be honest.

davisagli commented:
@prescod sorry but I'm still not clear on: what package do I need to install and what do I need to run to install it? I ran `make dev-install` and assumed that would install whatever needs to be installed.

prescod commented:
You need to install _snowfakery_. 

Plugins are registered on install.

`python setup.py install`

davisagli commented:
@prescod In CumulusCI, `make dev-install` includes `pip install -e .` (https://github.com/SFDO-Tooling/CumulusCI/blob/main/Makefile#L106) so I was assuming it was the same here. Any reason not to do that?

davisagli commented:
@prescod This is neat.

How are you planning to make use of the localized variations? They are not actually linked anywhere yet, right? Which makes the ~5 minutes required to generate them feel a bit much. Also, do you have a recommendation of where I can look to see proof that one of the localized variations is different? I poked around a bit at random and did not find obvious examples of localization.

prescod commented:
I fixed both of these issues today before I saw your comments. :) I changed the Makefile as you suggested.

The issue with not properly localizing was because a day ago I added another feature to allow me to customize the examples instead of totally auto-generating them.

I didn't hook the localizing code to the code that uses custom samples. I noticed this morning before my demo.

You can see French names and emails here:

* https://snowfakery.readthedocs.io/en/feature-fake-data-docs/fakedata/fr_FR.html#fake-email

There is a link to the localized versions here:

* https://snowfakery.readthedocs.io/en/feature-fake-data-docs/fakedata.html#localization

But it is probably too subtle. In a future PR I can make it more prominent.

jstvz reviewed (approved):
This is peak python, Paul. 

I have some concerns about how often we're running it in builds and warnings I'm seeing in the logs.

prescod commented on .github/workflows/ci.yml:
```diff
@@ -56,6 +56,25 @@ jobs:
+          python setup.py install
+
+      - name: Make Docs
+        run: make docs
```
Seems to warn on CI but in ReadTheDocs Pillow seems to be installed (which makes sense I guess).

https://snowfakery.readthedocs.io/en/feature-fake-data-docs/fakedata.html#fake-image

I hadn't even noticed that provider exists.

prescod commented on .github/workflows/ci.yml:
```diff
@@ -56,6 +56,25 @@ jobs:
+          make dev-install
+          python setup.py install
+
+      - name: Make Docs
```
Okay I'll think about that further.
